pghoard/basebackup_delta.py

```diff
@@ -0,0 +1,314 @@
+            "metadata": metadata,
+            "site": self.site,
+            "type": "UPLOAD",
+            "delta": {
+                "hexdigest": result_digest,
+            },
+        })
+
+        return input_size, result_size, result_digest
+
+    def _upload_files(self, callback_queue, todo, snapshotter):
+        todo_hexdigests = set(hash.hexdigest for hash in todo)
+
+        progress = Progress()
+        progress.add_total(len(todo_hexdigests))
+
+        def _submit_files_in_thread(hexdigest):
+            files = snapshotter.hexdigest_to_snapshotfiles.get(hexdigest, [])
+            for snapshotfile in files:
+                path = snapshotter.dst / snapshotfile.relative_path
+                if not path.is_file():
+                    self.log.error("%s disappeared post-snapshot", path)
+                    return progress.upload_missing
+                try:
+                    with snapshotfile.open_for_reading(snapshotter.dst) as f:
+                        size, stored_size, new_hash = self._upload_hexdigest_from_file(
+                            file_obj=f, relative_path=snapshotfile.relative_path, callback_queue=callback_queue
+                        )
+                        snapshotter.update_snapshot_file_data(
+                            relative_path=snapshotfile.relative_path,
+                            hexdigest=new_hash,
+                            file_size=size,
+                            stored_file_size=stored_size
+                        )
+                except Exception:  # pylint: disable=broad-except
+                    # Report failure - whole step will be retried later
+                    self.log.exception("Exception uploading %r", path)
+                    return progress.upload_failure
+
+            return progress.upload_success
+
+        def _result_cb(*, map_in, map_out):
+            # progress callback in 'main' thread
+            progress_callback = map_out
+            while True:
+                # Every thread is waiting for the transfer to finish, so we don't load disk too much,
+                # only `self.parallel` files in parallel
+                try:
+                    callback_queue.get(timeout=3.0)
+                    break
+                except Empty:
+                    continue
+
+            progress_callback(map_in)  # hexdigest
+            return True
+
+        sorted_todo_hexdigests = sorted(
+            todo_hexdigests, key=lambda hexdigest: -snapshotter.hexdigest_to_snapshotfiles[hexdigest][0].file_size
+        )
+        if not parallel_map_to(
+            fun=_submit_files_in_thread, iterable=sorted_todo_hexdigests, result_callback=_result_cb, n=self.parallel
+        ):
+            progress.add_fail()
+
+        progress.done()
+
+        if progress.finished_successfully:
+            self.log.info("All basebackup files were uploaded successfully")
+        else:
+            raise BackupFailure("Error while uploading backup files")
```

Comment:
When this fails how aggressively will it be retried? Could already be existing problem but just checking it won't end up in some very busy retry loop. It could be more relevant with this backup logic as this potentially causes a lot of calls to the cloud provider APIs while the regular tar backup mostly just loads the node locally.

Reply:
That's a good point, I guess it will try until succeeds, probably I need to add one metric, which should count how many backups failed and stop backup if more than X consecutive backups failed. What do you think?

Reply:
Maybe some exponential backoff so that it won't start new backup until enough time has passed since last failure?

Reply:
Actually infinite retrying of backup would happen in case if there is no backups, added retry/backoff there. For the usual backup request it won't happen.
However there is another retry/backoff mechanism in uploading part, I am not touching it, keeping the same logic.